proposals/0000-symbol-newtype-string.rst

```diff
@@ -0,0 +1,134 @@
+Make Symbol a newtype over String
+=================================
+
+.. author:: Oleg Grenrus (with amendments by Chris Dornan and Adam Gundry)
+.. date-accepted::
+.. ticket-url::
+.. implemented::
+.. highlight:: haskell
+.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/562>`_.
+.. sectnum::
+.. contents::
+
+I propose changing ``data Symbol`` definition
+to be
+
+::
+
+    newtype Symbol = MkSymbol String
+
+Motivation
+----------
+
+The motivation is similar to accepted and implemented
+`Unify Nat and Natural proposal <https://github.com/ghc-proposals/ghc-proposals/pull/364>`_.
+
+Consider the data type with fields of type ``String``::
+
+  data Fields = Fields [String]
+
+``Fields`` is inhabited by terms, and since recently (`The Char Kind <https://github.com/ghc-proposals/ghc-proposals/pull/387>`_) also by 
+types. However type literals produce ``Symbol``, not ``String`` ::
+
+  fs = Fields ["foo", "bar"]           -- ok
+  type FS = Fields '[ "foo", "bar" ]   -- not ok
+
+Alternatively, we could declare the type with fields of type ``Symbol``::
+
+  data Fields = Fields [Symbol]
+
+Then it would have the opposite issue::
+
+  fs = Fields ["foo", "bar" ]         -- not ok
+  type FS = Fields '[ "foo", "bar" ]  -- ok
+
+As with ``Nat`` and ``Natural`` proposal, it's possible to go roundabout
+this (and e.g. ``singletons-base`` does). However, if ``Symbol`` were
+an ordinary ``newtype`` over ``String``, we would avoid this issue entirely.
+
+(Note: ``singletons-base`` demotes ``Symbol`` to ``Text`` which is
+incorrect, as ``Text`` cannot represent all ``String`` values)
```

Comment:
Can't it? Do you have a reference for this claim? I'm infinitely curious now.

Reply:
From the haddock of `text`:

> A [Text](https://hackage.haskell.org/package/text-2.0.1/docs/Data-Text.html#t:Text) value is a sequence of Unicode scalar values, as defined in [§3.9, definition D76 of the Unicode 5.2 standard](http://www.unicode.org/versions/Unicode5.2.0/ch03.pdf#page=35). As such, a [Text](https://hackage.haskell.org/package/text-2.0.1/docs/Data-Text.html#t:Text) cannot contain values in the range U+D800 to U+DFFF inclusive. Haskell implementations admit all Unicode code points ([§3.4, definition D10](http://www.unicode.org/versions/Unicode5.2.0/ch03.pdf#page=13)) as [Char](https://hackage.haskell.org/package/base-4.14.1.0/docs/Data-Char.html#t:Char) values, including code points from this invalid range. This means that there are some [Char](https://hackage.haskell.org/package/base-4.14.1.0/docs/Data-Char.html#t:Char) values (corresponding to [Surrogate](https://hackage.haskell.org/package/base-4.14.1.0/docs/Data-Char.html#v:Surrogate) category) that are not valid Unicode scalar values, and the functions in this module must handle those cases.
> 
> Within this module, many functions construct a [Text](https://hackage.haskell.org/package/text-2.0.1/docs/Data-Text.html#t:Text) from one or more [Char](https://hackage.haskell.org/package/base-4.14.1.0/docs/Data-Char.html#t:Char) values. Those functions will substitute [Char](https://hackage.haskell.org/package/base-4.14.1.0/docs/Data-Char.html#t:Char) values that are not valid Unicode scalar values with the replacement character "�" (U+FFFD). Functions that perform this inspection and replacement are documented with the phrase "Performs replacement on invalid scalar values". The functions replace invalid scalar values, instead of dropping them, as a security measure. For details, see [Unicode Technical Report 36, §3.5](http://unicode.org/reports/tr36/#Deletion_of_Noncharacters).)

Reply:
Very nice.

That being said, is it relevant here? I don't think we can create `Symbol` values with surrogate code points in them, can we?

Reply:
I think we can?

```
ghci> :set -XDataKinds
ghci> type A = "\xd800"
ghci> :i A
type A :: GHC.Types.Symbol
type A = "\55296" :: GHC.Types.Symbol
```

Reply:
Just a tangentially related idea, but we can make `type Symbol = Text` if we do replacement on `Symbol` literals. `Text` is uninhabited at type level so that isn't a problem either. Indeed this is breaking, so it's not very desirable.

Reply:
Arfglbl. Indeed!

Though while `Symbol = Text` is technically breaking, it is probably not in a way that is depended upon in practice. So probably alright? Possibly maybe? But `Text` is not in base (yet?) so it's actually a pretty hard change to effect.

Reply:
> But Text is not in base (yet?)

It shouldn't be hard to move the type declaration to `base` (`GHC.Types.Text` maybe?) but keep all the combinators for working on it in `text`. Just move these few lines of code:

```
data Text = Text
    {-# UNPACK #-} !Array -- bytearray encoded as UTF-8
    {-# UNPACK #-} !Int     -- offset in bytes (not in Char!), pointing to a start of UTF-8 sequence
    {-# UNPACK #-} !Int     -- length in bytes (not in Char!), pointing to an end of UTF-8 sequence

data Array = ByteArray ByteArray#
```

Reply:
It wouldn't even need to be `base`; `ghc-prim` is enough. `text` also depends on `ghc-prim` already, so all of this can be done without CLC.